### PR TITLE
Addition of modularMultiply, gcd, and exactDivide

### PR DESCRIPTION
**Pull request by jacobwil**
Hello! I'm a contributor on [Apache Pirk](http://pirk.incubator.apache.org). We have been using jna-gmp to accelerate our modPow operations for a few months. In an effort to improve performance of [our implementation](https://github.com/apache/incubator-pirk/blob/master/src/main/java/org/apache/pirk/encryption/Paillier.java) of the [Paillier encryption scheme](https://en.wikipedia.org/wiki/Paillier_cryptosystem) I have exposed modularMultiply, gcd, and exactDivide functionality from GMP. You can see my unmerged Pirk code in my fork of the Pirk codebase [here](https://github.com/jacobwil/incubator-pirk/blob/JNAGMPForModularMultiply/src/main/java/org/apache/pirk/encryption/IntegerMathAbstraction.java) and [here](https://github.com/jacobwil/incubator-pirk/blob/JNAGMPForModularMultiply/src/main/java/org/apache/pirk/encryption/Paillier.java).

I submitted the ICLA form a few minutes ago. 

I have been (micro)benchmarking what the use of GMP does to our integer math performance using JMH (simulating what the numbers we work with usually look like) and here are the results. Benchmark code is [here](https://github.com/jacobwil/incubator-pirk/tree/JNAGMPForModularMultiply/src/main/java/org/apache/pirk/benchmark)

    Benchmark                                   Mode  Cnt        Score        Error  Units
    ExactDivideBenchmark.testWithGMP           thrpt  200   110961.159 ±   1071.711  ops/s
    ExactDivideBenchmark.testWithoutGMP        thrpt  200    53641.770 ±    267.455  ops/s
    GCDBenchmark.testWithGMP                   thrpt  200    24121.188 ±    144.105  ops/s
    GCDBenchmark.testWithoutGMP                thrpt  200     1243.952 ±      7.829  ops/s
    ModBenchmark.testWithGMP                   thrpt  200    93148.587 ±    762.959  ops/s
    ModBenchmark.testWithoutGMP                thrpt  200  2300046.853 ± 267769.213  ops/s
    ModularInverseBenchmark.testWithGMP        thrpt  200     9891.903 ±     38.648  ops/s
    ModularInverseBenchmark.testWithoutGMP     thrpt  200      390.049 ±      1.966  ops/s
    ModularMultiplyBenchmark.testWithGMP       thrpt  200    36154.112 ±    219.346  ops/s
    ModularMultiplyBenchmark.testWithoutGMP    thrpt  200    15777.195 ±    100.081  ops/s
    MultiplyBenchmark.testWithGMP              thrpt  200   102739.040 ±    522.167  ops/s
    MultiplyBenchmark.testWithoutGMP           thrpt  200   229137.717 ±    658.034  ops/s
    PaillierBenchmark.testWithGMP              thrpt  200       26.943 ±      0.122  ops/s
    PaillierBenchmark.testWithGMPConstantTime  thrpt  200       19.695 ±      0.078  ops/s
    PaillierBenchmark.testWithoutGMP           thrpt  200        4.573 ±      0.012  ops/s

*Note: PaillierBenchmark is chiefly a modPow benchmark*

Takeaways: 
 - Modular multiply constructed from GMP mul and mod produces a 1.3x speedup over BigInteger mul and mod
 - Mod and Mul on their own each appear to be slower to run in GMP through JNA versus just using BigInteger
 - In our application gcd is *more than an order of magnitude faster* when using jnagmp (so is modInverse)

If my changes are accepted we would *greatly* appreciate a new release containing these features so we can include the new version through our pom file. That way other users and developers can benefit from these performance increases.

### CODE REVIEW

**Comment by dragonsinth** [comment]
Hi Jacob, thanks for the contribution!

@JakeWharton did you get the ICLA?

I'll take a look this week.  Do you have a theory about why mul and mod would be slower individually, but faster when composited?  Is the problem that the speed is dominated by translating the raw data between Java and native?

It'd be interesting to see if using a GmpIntegers with cached peers over a series of operations to avoid the import / export would be faster than the Java implementation.

**Comment by jacobwil** [comment]
Scott, my theory on the mul and mod performance is exactly the same as your intuition: that those operations are dominated by the time required to convert data and call into GMP using JNA. 
Or maybe I designed my benchmarks poorly (in which case it may be a good idea to make a new module of this project for some microbenchmarks implemented correctly for future additions). 

After seeing those results above and disbelieving them I wrote [expanded modular multiplication benchmarks](https://github.com/jacobwil/incubator-pirk/blob/JNAGMPForModularMultiply/src/main/java/org/apache/pirk/benchmark/ModularMultiplyBenchmark.java). I didn't expect to find that using half GMP and half BigInteger would be more efficient but now I can say I tried it. 

    Benchmark                                              Mode  Cnt      Score     Error  Units
    ModularMultiplyBenchmark.testWithBigIntegerMulGMPMod  thrpt  120  27344.445 ±  69.231  ops/s
    ModularMultiplyBenchmark.testWithGMP                  thrpt  120  35632.620 ± 214.907  ops/s
    ModularMultiplyBenchmark.testWithGMPMulBigIntegerMod  thrpt  120  15500.034 ± 147.148  ops/s
    ModularMultiplyBenchmark.testWithGMPSeparateCalls     thrpt  120  26046.791 ± 327.106  ops/s
    ModularMultiplyBenchmark.testWithoutGMP               thrpt  120  16275.952 ±  52.940  ops/s

Can you expand on your thinking with this quote?
> It'd be interesting to see if using a GmpIntegers with cached peers over a series of operations to avoid the import / export would be faster than the Java implementation.

By multiple operations do you mean between the mul and mod in [the implementation of modularMultiply](https://github.com/jacobwil/jna-gmp/blob/pirkAdditions/jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java#L366)? If I understand how the mpz_t objects and JNA work then I'm already doing that.  
Or do you mean between multiple implications of modularMultiply (and others)? Perhaps by checking in [GMP.getPeer](https://github.com/jacobwil/jna-gmp/blob/pirkAdditions/jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java#L406) to see if the `value` was the last thing loaded into `sharedPeer` and that `sharedPeer`'s value hasn't changed.

**Comment by JakeWharton** [comment]
Yes. The CLA is signed.

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 1]
```diff
@@ -162,6 +169,78 @@ public static BigInteger modInverse(BigInteger val, BigInteger modulus) {
...
+   * Divide dividend by divisor. This method only returns correct answers when the division produces
+   * no remainder. Correct answers should not be expected when the divison would result in a
+   * remainder.
+   *
+   * @param dividend
+   * @param divisor
+   * @return dividend / divisor
+   */
+  public static BigInteger exactDivide(BigInteger dividend, BigInteger divisor) {
+    return INSTANCE.get().exactDivImpl(dividend, divisor);
```
Comment (dragonsinth) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: I think we need to check for divisor == 0 here to avoid a native exception

**Comment by dragonsinth** [comment]
Adding gcd and exactDivide seem very natural to me.  I'm a little less convinced about the rest.  Exposing `mul` and `mod` in particular seems like kind of a trap.

mulMod seems interesting, I could see it.  But I think I'd like to see a benchmark comparing a few implementations of modular multiply:

1. `(A * B) mod C` java
2. `(A * B) mod C` native
3. `(A mod C * B mod C) mod C` java
4. `(A mod C * B mod C) mod C` native, with maximal peer reuse

You've done the first two, I'd like to see the last two.  I could imagine the third case actually winning for practical applications.

**Comment by jacobwil** [comment]
I can do an experimental implementation of the other native function and benchmark it with the same existing class I have in Pirk (I won't check that into our codebase but I'll attach the file to the followup comment when I have results). 

I'm up for dropping mod and mul—they're definitely traps. Wrote them first as I was finding my way around the project. Do we want to consider switching them to protected or move them into the `GmpTests` class so we can continue to directly test the behavior of the underlying JNA-exposed GMP mod and mul functions?

And I'll add the divisor == 0 check to exactDivide. That could have been unfortunate.

**Comment by dragonsinth** [comment]
Sounds good.

I think the right thing to do here is just expose `mod` and `mul` only as LibGmp natives, and you can just add direct tests in LibGmpTest.  If you do that, you should be able to optimize your own code paths even more just by chaining together a series of native calls that will outperform anything we could add to Gmp.

**Comment by jacobwil** [comment]
I have returned with benchmark results: 

    Benchmark                                                Mode  Cnt      Score     Error  Units
    ModularMultiplyBenchmark.testWithGMPMulMod              thrpt  200  34053.772 ± 828.238  ops/s
    ModularMultiplyBenchmark.testWithGMPMulModMulModMod     thrpt  200  33004.568 ± 621.534  ops/s
    ModularMultiplyBenchmark.testWithoutGMP                 thrpt  200  13345.274 ± 372.364  ops/s
    ModularMultiplyBenchmark.testWithoutGMPMulModMulModMod  thrpt  200  13382.723 ± 180.021  ops/s

I posted a [new repository with just the benchmarks](https://github.com/jacobwil/JMH-JNAGMP). They're mostly ripped out of Apache Pirk. Right now the pom file pulls a JAR of my branch of JNAGMP out of my github repo (so it'll break as soon as I remove the `mulModImplAlt`). 

`testWithGMPMulMod`, which is the original `A*B mod C` version using GMP, is still the fastest. 

Here was my implementation deep down of the alternative modular multiplication ([see in context in this version of GMP.java](https://gist.github.com/jacobwil/cba3a425f7a3b735bb39a554c183375e))

```java
private BigInteger mulModImplAlt(BigInteger factor1, BigInteger factor2, BigInteger mod) {
  // (A mod C * B mod C) mod C
  //
  // A -> sharedOperands[0]
  // B -> sharedOperands[1]
  // C -> sharedOperands[2]
  //
  // A mod C -> sharedOperands[3]
  // B mod C -> sharedOperands[0]
  // (A mod C) * (B mod C) -> sharedOperands[1]
  // ((A mod C) * (B mod C)) mod C -> sharedOperands[0]
  mpz_t factor1Peer = getPeer(factor1, sharedOperands[0]);
  mpz_t factor2Peer = getPeer(factor2, sharedOperands[1]);
  mpz_t modPeer = getPeer(mod, sharedOperands[2]);

  __gmpz_mod(sharedOperands[3], factor1Peer, modPeer);
  __gmpz_mod(sharedOperands[0], factor2Peer, modPeer);

  __gmpz_mul(sharedOperands[1], sharedOperands[3], sharedOperands[0]);

  __gmpz_mod(sharedOperands[0], sharedOperands[1], modPeer);

  // The result size should be <= mod size, but round up to the nearest byte.
  int requiredSize = (mod.bitLength() + 7) / 8;
  return new BigInteger(mpzSgn(sharedOperands[0]), mpzExport(sharedOperands[0], requiredSize));
}
```

**Comment by dragonsinth** [comment]
Awesome work.  I was able to repeat the benchmarks, and the MulModMulModMod formulation only seems to help if you know up front that the operands are much bigger than the modulus-- if I removed the `.pow(2)` on the modulus from ModularMultiplyBenchmark, then the alt formulation was significantly faster.  But if the modulus is same size or bigger it doesn't help obviously.

So I think want I want to do is pull in gcd and exact divide to the Gmp layer, and just add mul and mod to the LibGmp layer, which should enable you to just call that layer directly in your code.  Sound good?

**Comment by jacobwil** [comment]
That's a good observation on the length of multiplicand factors vs modulus. I'm going to see if I can reason out a way to predict which one will be faster using the BigInteger's bitlength. I would bet that there's a way to leverage JMH to generate experiments for me—could be a good learning experience. 

If there turns out to be a good way to do it then we can just leave both internal implementations in place (with better naming and copious explanation) and have `Gmp.modularMultiply` pick which implementation to use.

**Comment by jacobwil** [comment]
It turns out that JMH is **awesome** and made this shockingly easy to implement. (Seriously, take a look at [this file](https://github.com/jacobwil/JMH-JNAGMP/blob/master/src/main/java/org/jacobwilder/JMHJNAGMP/ModularMultiplyBenchmarkParamSpace.java#L27), that `@Param({…})` specification was all it took)

Here are the results that I have. The two Score columns are in ops/sec. The speedup column is the faster score divided by the slower score minus one rounded to two decimal places. 
```
-------------  --------------  ---------------  ------------------------  ---------------  -------
(FACTOR_SIZE)  (MODULUS_SIZE)  Score GMPMulMod  Score GMPMulModMulModMod  Faster Version   Speedup
1024           1024            135219.089       127481.098                MulMod           0.06
1024           2048            119314.118       112797.858                MulMod           0.06
1024           3074            106926.6         101447.4                  MulMod           0.05
1024           4096            94872.184        89335.507                 MulMod           0.06
1024           8192            66856.722        65039.01                  MulMod           0.03
2048           1024            92378.083        89010.837                 MulMod           0.04
2048           2048            83649.035        82110.773                 MulMod           0.02
2048           3074            77371.731        75673.047                 MulMod           0.02
2048           4096            70661.721        69994.128                 MulMod           0.01
2048           8192            55613.747        51950.81                  MulMod           0.07
3074           1024            68620.831        69878.844                 MulModMulModMod  0.02
3074           2048            59967.654        58274.661                 MulMod           0.03
3074           3074            60307.771        58134.988                 MulMod           0.04
3074           4096            57112.215        55486.284                 MulMod           0.03
3074           8192            45467.956        44934.575                 MulMod           0.01
4096           1024            53837.839        59091.77                  MulModMulModMod  0.10
4096           2048            47555.603        45500.227                 MulMod           0.05
4096           3074            44543.357        43886.02                  MulMod           0.01
4096           4096            45974.204        45731.662                 MulMod           0.01
4096           8192            38885.574        38738.002                 MulMod           0.00
8192           1024            29376.831        36167.688                 MulModMulModMod  0.23
8192           2048            24613.781        27622.644                 MulModMulModMod  0.12
8192           3074            21892.998        23043.419                 MulModMulModMod  0.05
8192           4096            21286.526        20824.538                 MulMod           0.02
8192           8192            24008.546        23446.881                 MulMod           0.02
-------------  --------------  ---------------  ------------------------  ---------------  -------
```

From this data I see that the difference between `MulMod` and `MulModMulModMod` really only shows up significantly when `FACTOR_SIZE

**Comment by jacobwil** [comment]
I just pushed an addition with a unified `Gmp.modularMultiply` that selects what is likely to be the fastest implementation of mulMod internally. 

I ran a huge amount of benchmarking and found that the `A * B mod C` version is the faster version up until the point where the max bit length of the product of the multiplicands is twice the bit length of the modulus. After that point, `((A mod C) * (B mod C)) mod C` starts being equal or faster. 

Below is the data I got from running [my benchmark](https://github.com/jacobwil/JMH-JNAGMP/blob/febce7dd171a529f961260ab19a4af2632b86318/src/main/java/org/jacobwilder/JMHJNAGMP/ModularMultiplyBenchmarkParamSpace.java) (which no longer has the `.pow(2)`)  with the default JMH options (10 forks with 20 measurements of the microbenchmark each):
```
-------------  --------------  ------------------------  --------------------------  ---------------  -------  ----------------------------
(FACTOR_SIZE)  (MODULUS_SIZE)  Score GMPMulModMulModMod  testWithGMPMulModMulModMod  faster version   speedup  expected multiplication 
bits           bits            ops/sec                   ops/sec                                               product len over modulus len
1024           32768           41941.576                 41527.115                   MulMod           0.01     0.06
1024           16384           67428.651                 64786.942                   MulMod           0.04     0.12
2048           32768           37033.564                 36203.209                   MulMod           0.02     0.12
3074           32768           32790.989                 31866.523                   MulMod           0.03     0.19
1024           8192            93941.355                 90926.006                   MulMod           0.03     0.25
2048           16384           54630.308                 53635.046                   MulMod           0.02     0.25
4096           32768           29007.709                 28657.395                   MulMod           0.01     0.25
3074           16384           45605.893                 44388.948                   MulMod           0.03     0.38
1024           4096            119379.610                111548.264                  MulMod           0.07     0.5
2048           8192            72241.943                 69919.974                   MulMod           0.03     0.5
4096           16384           39266.409                 38305.214                   MulMod           0.03     0.5
8192           32768           19873.991                 19641.453                   MulMod           0.01     0.5
1024           3074            127890.224                119251.063                  MulMod           0.07     0.67
3074           8192            56798.926                 55329.989                   MulMod           0.03     0.75
1024           2048            134066.870                126961.812                  MulMod           0.06     1.0
2048           4096            84579.632                 81581.376                   MulMod           0.04     1.0
4096           8192            46434.568                 46043.008                   MulMod           0.01     1.0
8192           16384           23976.375                 23598.790                   MulMod           0.02     1.0
16384          32768           11428.207                 11276.457                   MulMod           0.01     1.0
2048           3074            86096.379                 83029.073                   MulMod           0.04     1.33
3074           4096            60157.268                 59115.015                   MulMod           0.02     1.5
1024           1024            151495.681                137893.715                  MulMod           0.1      2.0
2048           2048            92596.333                 87856.624                   MulMod           0.05     2.0
3074           3074            62421.144                 59499.029                   MulMod           0.05     2.0
4096           4096            47555.038                 46153.272                   MulMod           0.03     2.0
8192           8192            21295.347                 20546.030                   MulMod           0.04     2.0
16384          16384           8759.199                  8677.818                    MulMod           0.01     2.0
4096           3074            49148.087                 50741.902                   MulModMulModMod  0.03     2.66
3074           2048            68501.233                 69767.202                   MulModMulModMod  0.02     3.0
2048           1024            106785.567                106291.032                  MulMod           0.0      4.0
4096           2048            54555.372                 59169.581                   MulModMulModMod  0.08     4.0
8192           4096            24688.721                 27737.330                   MulModMulModMod  0.12     4.0
16384          8192            9784.031                  11393.406                   MulModMulModMod  0.16     4.0
8192           3074            25916.515                 30147.842                   MulModMulModMod  0.16     5.33
3074           1024            79956.691                 85551.973                   MulModMulModMod  0.07     6.0
4096           1024            64130.189                 73796.897                   MulModMulModMod  0.15     8.0
8192           2048            29308.312                 36004.528                   MulModMulModMod  0.23     8.0
16384          4096            11693.253                 15341.934                   MulModMulModMod  0.31     8.0
16384          3074            12436.804                 16883.914                   MulModMulModMod  0.36     10.66
8192           1024            34202.027                 45474.618                   MulModMulModMod  0.33     16.0
16384          2048            13816.158                 20222.440                   MulModMulModMod  0.46     16.0
16384          1024            16077.462                 25434.625                   MulModMulModMod  0.58     32.0
-------------  --------------  ------------------------  --------------------------  ---------------  -------  ----------------------------
```

I'll note that as it stands my [JMH-JNAGMP](https://github.com/jacobwil/JMH-JNAGMP) no longer works because the alternate MulModMulModMod method `Gmp.modularMultiplyAlt` it relies on doesn't exist in my newest version. 

I'll take care of removing the external accessibility of `mod` and `pow` now.

**Comment by dragonsinth** [comment]
@jacobwil I pushed commits to master to cover gcd, exactDivide, and addition of the native mul and mod stubs.  I'm not convinced that mulMod is in scope for this project.  The project goals are to expose the underlying GMP functions to make it safe and easy to build fast math utils.  I think we're in agreement that providing porcelain for mul and mod is not helpful since naively using them is not faster.  mulMod is useful, but it seems to do too much to fit the project scope.  If mulMod, why not addMod, etc?  The list of chained operations you might want to do that could be really fast if you can avoid multiple import/exports just goes on and on.

That said, with the current state of master, and mul and mod exposed, you should be able to just put mulMod in your project.

One last note: I'm be completely open to ideas about how to make it easier to use the native functions more directly.  LibGmpTest is kind of a train wreck because it's so difficult to work with the native functions.  If you have some ideas for utility code or types that would make LibGmpTest much easier to write, and would allow you to implement mulMod easier, that would be great.

**Pull request closed without merging**